### PR TITLE
don't create schema for terms that already have them

### PR DESCRIPTION
**Pull request by kleinschmidt**
possible fix for #97: don't generate schema entries for terms that don't need schema setting (so categorical, continuous, intercept).

The alternative is to specify `concrete_term` method like `concrete_term(t, d, hint) = t` that just passes anything through that's not a `Term`.  I'm pretty sure these are functionally basically equivalent, but maybe not.

@oxinabox 

EDIT: I also added a reproducer and fix for #95, which required a wrapper type `Schema` in order to avoid method ambiguities in `apply_schema`.

### CODE REVIEW

**Comment by nalimilan** [comment]
Is there a reason why we can't just use `Dict` instead of defining a custom `Schema` type?

**Comment by kleinschmidt** [comment]
Not really, although having a wrapper type allows for more convenience constructors (e.g., if you use symbols as keys in your constructor instead of `Term`s).  But really I just added it for dispatch so any type would work for that.

**Comment by nalimilan** [comment]
> Not really, although having a wrapper type allows for more convenience constructors (e.g., if you use symbols as keys in your constructor instead of `Term`s).

But aren't symbols accepted already in dicts? That can be handled when applying the schema.

**Comment by kleinschmidt** [comment]
At the moment you can pass a dict with symbol keys and it'll choke since `apply_schema` tries to use `Term`s to index.  So the way that happens will have to be changed in every single `apply_schema` method (and worse, anyone who writes their own method will have to be careful to include whatever support code is necessary).  That sounds like too much boilerplate to me, which can be avoided by using a wrapper type that provides a single place for enforcing a particular standard and accessing the elements.

Anyway, I'm not really wedded to using a wrapper type (and in fact I tried to avoid it originally, using a vanilla dict), but having something concrete that's easy to dispatch on makes it much easier to avoid method ambiguities which were becoming a big headache for `apply_schema`.

File: src/schema.jl [hunk 1]
```diff
@@ -239,14 +283,34 @@ end
 # to know whether to repair, need to know context a term appears in.  main
 # effects occur in "own" context.
 
+"""
+    apply_schema(t::AbstractTerm, schema::StatsModels.FullRank, Mod::Type)
+
+Apply a schema, under the assumption that when a less-than-full rank model
+matrix would be produce, categorical terms should be "promoted" to full rank
```
Comment (nalimilan) [inline src/schema.jl]:
```suggestion
matrix would be produced, categorical terms should be "promoted" to full rank
```

**Comment by nalimilan** [comment]
OK. One solution would be to have terms and symbols compare and hash equal (so that they can be used interchangeably in dicts), but I guess you wouldn't want that.

File: src/schema.jl [hunk 2]
```diff
@@ -18,6 +18,41 @@ terms(t::TupleTerm) = mapreduce(terms, union, t)
...
+
+Struct that wraps a `Dict` mapping `Term`s to their concrete forms.  This exists
+mainly for dispatch purposes and to support possibly more sophisticated behavior
+in the future.
+
+A `Schema` behaves for all intents and purposes like an immutable `Dict`, and
+delegates the constructor as well as `getindex`, `get`, `merge!`, `merge`,
+`keys`, and `haskey` to the wrapped `Dict`.
+"""
+struct Schema
```
[the remark this answers is not in the document]
Comment (kleinschmidt) [reply src/schema.jl]: Is there any advantage to subtyping like this?  IIRC I didn't implement `setindex!` so this doens't really implement the full `AbstractDict` interface, right?

**Pull request merged**